skrifa/src/color/mod.rs

```diff
@@ -1,6 +1,406 @@
+    types::{BoundingBox, GlyphId, Point},
+    ReadError, TableProvider,
+};
+
+use std::{collections::HashSet, fmt::Debug, ops::Range};
+
+use colrv1_traversal::{get_clipbox_font_units, traverse_with_callbacks};
+
+pub use transforms::Transform;
+
+use crate::prelude::LocationRef;
+
+use self::instance::{resolve_paint, PaintId};
+
+/// An error during drawing a COLR glyph. This covers inconsistencies
+/// in the COLRv1 paint graph as well as downstream
+/// parse errors from read-fonts.
+#[derive(Debug, Clone)]
+pub enum ColorError {
+    ParseError(ReadError),
+    NoColrV1GlyphForId(GlyphId),
+    PaintCycleDetected,
+}
+
+impl std::fmt::Display for ColorError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ColorError::ParseError(read_error) => {
+                write!(f, "Error parsing font data: {read_error}")
+            }
+            ColorError::NoColrV1GlyphForId(glyph_id) => {
+                write!(f, "No COLRv1 glyph found for glyph id: {glyph_id}")
+            }
+            ColorError::PaintCycleDetected => write!(f, "Paint cycle detected in COLRv1 glyph."),
+        }
+    }
+}
+
+impl From<ReadError> for ColorError {
+    fn from(value: ReadError) -> Self {
+        ColorError::ParseError(value)
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+// This repr(C) is required so that C-side FFI's
+// are able to cast the ColorStop slice to a C-side array pointer.
+#[repr(C)]
+#[derive(PartialEq)]
+#[cfg_attr(test, derive(Serialize, Deserialize))]
+/// A color stop of a gradient. All gradient callbacks of [`ColorPainter`] normalize color stops to be in the range of 0
+/// to 1. `palette_index` specifies a color from the CPAL table, to be multiplied with `alpha` before use.
+pub struct ColorStop {
+    pub offset: f32,
+    pub palette_index: u16,
+    pub alpha: f32,
+}
+
+// Design considerations for choosing a slice of ColorStops as `color_stop`
+// type: In principle, a local `Vec<ColorStop>` allocation would not required if
+// we're willing to walk the `ResolvedColorStop` iterator to find the minimum
+// and maximum color stops.  Then we could scale the color stops based on the
+// minimum and maximum. But performing the min/max search would require
+// re-applying the deltas at least once, after which we would pass the scaled
+// stops to client side and have the client sort the collected items once
+// again. If we do want to pre-ort them, and still use use an
+// `Iterator<Item=ColorStop>`` instead as the `color_stops` field, then we would
+// need a Fontations-side allocations to sort, and an extra allocation on the
+// client side to `.collect()` from the provided iterator before passing it to
+// drawing API.
+//
+/// Used for encapsulating the different types of fills that may occur in a
+/// COLRv1 glyph. The client receives the information about the fill type in the
+/// [`fill``](ColorPainter::fill) callback of the [`ColorPainter`] trait.
```

Comment:
Same thing with separating first sentence into it's own line.

Reply:
Done.